src/core/thread/network_data_leader.cpp

```diff
@@ -531,6 +549,291 @@ ThreadError Leader::RemoveCommissioningData(void)
+{
+    ThreadError error = kThreadError_NotFound;
+    Dhcp6AlocRlocCache *prevCache = NULL;
+    Dhcp6AlocRlocCache *cache = NULL;
+
+    VerifyOrExit(mDhcp6AlocRlocCacheHead, ;);
+    cache = mDhcp6AlocRlocCacheHead;
+
+    while (cache)
+    {
+        if (cache->IsEntry(aContextId, aRloc16))
+        {
+            if (!prevCache)
+            {
+                mDhcp6AlocRlocCacheHead = cache->GetNext();
+            }
+            else
+            {
+                prevCache->SetNext(cache->GetNext());
+            }
+
+            cache->SetNext(mDhcp6AlocRlocCacheAvail);
+            mDhcp6AlocRlocCacheAvail = cache;
+            ExitNow(error = kThreadError_None);
+        }
+
+        prevCache = cache;
+        cache = cache->GetNext();
+    }
+
+exit:
+    return error;
+}
+
+ThreadError Leader::GetDhcp6AlocRlocCacheEntrybyContextId(uint8_t aContextId, uint16_t &aRloc16)
+{
+    ThreadError error = kThreadError_NotFound;
+    Dhcp6AlocRlocCache *cache = NULL;
+
+    VerifyOrExit(mDhcp6AlocRlocCacheHead, ;);
+    cache = mDhcp6AlocRlocCacheHead;
+
+    while (cache)
+    {
+        if (cache->GetContextId() == aContextId)
+        {
+            aRloc16 = cache->GetRloc16();
+            ExitNow(error = kThreadError_None);
+        }
+
+        cache = cache->GetNext();
+    }
+
+exit:
+    return error;
+}
+
+bool Leader::UpdateDhcp6AlocRlocCacheEntries()
+{
+    otNetworkDataIterator iterator = OT_NETWORK_DATA_ITERATOR_INIT;
+    otBorderRouterConfig config;
+    bool found = false;
+    bool rval = false;
+
+    Dhcp6AlocRlocCache *cache = mDhcp6AlocRlocCacheHead;
+
+    while (cache)
+    {
+        found = false;
+
+        while (GetNextOnMeshPrefix(&iterator, &config) == kThreadError_None)
+        {
+            if (!config.mDhcp)
+            {
+                continue;
+            }
+
+            if (cache->IsEntry(config.mContextId, config.mRloc16))
+            {
+                found = true;
+                break;
+            }
+
+            if (cache->GetContextId() == config.mContextId)
+            {
+                cache->SetRloc16(config.mRloc16);
+                found = true;
+                break;
+            }
+        }
+
+        if (!found)
+        {
+            rval = true;
+        }
+
+        cache = cache->GetNext();
+    }
+
+    while (GetNextOnMeshPrefix(&iterator, &config) == kThreadError_None)
+    {
+        if (!config.mDhcp)
+        {
+            continue;
+        }
+
+        found = false;
+        cache = mDhcp6AlocRlocCacheHead;
+
+        while (cache)
+        {
+            if (cache->IsEntry(config.mContextId, config.mRloc16))
+            {
+                found = true;
+                break;
+            }
+
+            cache = cache->GetNext();
+        }
+
+        if (!found)
+        {
+            rval = true;
+            VerifyOrExit((AddDhcp6AlocRlocCacheEntry(config.mContextId, config.mRloc16) == kThreadError_None), ;);
```

Comment:
Thanks for the explanation.  I understand what you're trying to do.  At the same time, I'm not sure a Thread device should send a DHCPv6 Release message in response to the DHCPv6 Agent being removed from the Network Data.  In particular, the DHCPv6 Agent may not even be there to receive the Release message.  Maybe in this case, the Thread device should just clean up its state (i.e. remove the address and DHCP Client state).  Is there a reason to send the Release message?